File: Ogury/SASOguryBaseAdapter.m
//
//  SASOguryBaseAdapter.m
//  AdViewer
//
//  Created by Loïc GIRON DIT METAZ on 30/09/2020.
//  Copyright © 2020 Smart AdServer. All rights reserved.
//

#import "SASOguryBaseAdapter.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SASOguryBaseAdapter

- (void)configureOgurySDKWithServerParameterString:(NSString *)serverParameterString
                               andClientParameters:(NSDictionary *)clientParameters
                                 completionHandler:(void(^)(NSError * _Nullable))completionHandler {
    // Retrieve the asset key and the ad unit id
    NSArray *serverParameters = [serverParameterString componentsSeparatedByString:@"|"];
    
    // Invalid parameter string, the loading will be cancelled with an error
    if (serverParameters.count != 2 && serverParameters.count != 3 && serverParameters.count != 6) {
        NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeInvalidParameterString userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Invalid server parameter string: %@", serverParameterString] }];
        completionHandler(error);
        return;
    }
    
    // Parsing generic parameters
    NSString *assetKey = serverParameters[0];
    self.adUnitId = serverParameters[1];
    
    // Banner parameters
    if (serverParameters.count == 3) {
        // Invalid parameter string, the loading will be cancelled with an error
        if (![serverParameters[2] respondsToSelector:@selector(integerValue)]) {
            NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeInvalidParameterString userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Invalid server parameter string: %@", serverParameterString] }];
            completionHandler(error);
            return;
        }
        
        // Setting banner size
        if ([serverParameters[2] integerValue] == 0) {
            self.bannerSize = OguryAdsBannerSize.small_banner_320x50;
        } else if ([serverParameters[2] integerValue] == 1) {
            self.bannerSize = OguryAdsBannerSize.mpu_300x250;
        }
    }
    
    // Thumbnail parameters
    if (serverParameters.count == 6) {
        // Invalid parameter string, the loading will be cancelled with an error
        if (![serverParameters[2] respondsToSelector:@selector(integerValue)]
            || ![serverParameters[3] respondsToSelector:@selector(integerValue)]
            || ![serverParameters[4] respondsToSelector:@selector(integerValue)]
            || ![serverParameters[5] respondsToSelector:@selector(integerValue)]) {
            NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeInvalidParameterString userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Invalid server parameter string: %@", serverParameterString] }];
            completionHandler(error);
            return;
        }
        
        // Setting Thumbnail size
        SASOguryAdapterThumbnailSize size;
        size.maxWidth = [serverParameters[2] integerValue];
        size.maxHeight = [serverParameters[3] integerValue];
        size.leftMargin = [serverParameters[4] integerValue];
        size.topMargin = [serverParameters[5] integerValue];
        
        self.thumbnailSize = size;
    }
 
    // Retrieving IAB TCF Consent String v2 from shared preferences
    NSString *tcfString = [[NSUserDefaults standardUserDefaults] objectForKey:@"IABTCF_TCString"];
    if (tcfString != nil) {
        // Setting the TCF string on the Ogury SDK
        [OguryChoiceManagerExternal setConsentForTCFV2WithAssetKey:assetKey iabString:tcfString andNonIABVendorsAccepted:@[]];
    }
    
    // Initializing Ogury SDK
    [[OguryAds shared] setupWithAssetKey:assetKey andCompletionHandler:^(NSError *error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (error == nil) {
                // Ogury has been successfully initialized
                completionHandler(nil);
            } else {
                // An error occurred when initializing Ogury
                NSError *handlerError = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeCannotInitializeOgurySDK userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Ogury SDK returned error: %@", error] }];
                completionHandler(handlerError);
            }
        });
    }];
}

@end

NS_ASSUME_NONNULL_END

File: Ogury/SASOguryBaseAdapter.swift
//
//  SASOguryBaseAdapter.swift
//  SmartAdServer
//
//  Created by Loïc GIRON DIT METAZ on 21/07/2021.
//  Copyright © 2021 Smart AdServer. All rights reserved.
//

import UIKit
import SASDisplayKit
import OguryAds
import OgurySdk

/**
 Ogury base adapter class for Smart SDK mediation.
 
 Mediation adapter classes can be used to display an ad using a third party SDK directly from
 an insertion handled by Smart.
 
 To use an adapter class, you simply have to add them to your Xcode project and they will
 be automatically instantiated by the Smart SDK if needed.
 */
class SASOguryBaseAdapter: NSObject {
    
    /**
     Constants used to produce errors.
     */
    class ErrorConstants {
        static let errorDomain = "SASOguryAdapter"
        
        static let errorCodeInvalidParameterString = 1
        static let errorCodeAdNotAvailable = 2
        static let errorCodeAdNotLoaded = 3
        static let errorCodeAdError = 4
        static let errorCodeCannotInitializeOgurySDK = 5
        
        static let oguryNoAdErrorCode = 2008
    }
    
    /**
     Hold the different sizes needed to display an Ogury Thumbnail.
     */
    struct ThumbnailSize {
        let maxWidth: Int
        let maxHeight: Int
        let leftMargin: Int
        let topMargin: Int
    }
    
    /// Ogury Ad Unit ID.
    var adUnitId: String? = nil
    
    /// Ogury Thumbnail sizes.
    var thumbnailSize: ThumbnailSize? = nil
    
    /// Ogury Banner size.
    var bannerSize: OguryAdsBannerSize? = nil
    
    /**
     Configure the Asset Key and the Ad Unit Id used by Ogury from the server parameter string provided
     by the Smart SDK.
     
     @param serverParameterString The server parameter string provided by Smart.
     @param clientParameters      The client parameters dictionary provided by Smart.
     @param completion            Handler called when the Ogury SDK setup is finished (it will return an error in case of failure, nil if successful).
     */
    func configureOgurySDK(serverParameterString: String, clientParameters: [AnyHashable: Any], completion: @escaping (Error?) -> ()) {
        // Retrieve the asset key and the ad unit id
        let serverParameters = serverParameterString.split(separator: "|")
        
        // Invalid parameter string, the loading will be cancelled with an error
        guard serverParameters.count == 2 || serverParameters.count == 3 || serverParameters.count == 6 else {
            completion(NSError(
                domain: ErrorConstants.errorDomain,
                code: ErrorConstants.errorCodeInvalidParameterString,
                userInfo: [NSLocalizedDescriptionKey: "Invalid server parameter string: \(serverParameters)"]
            ))
            return
        }
        
        // Parsing generic parameters
        let assetKey = String(serverParameters[0])
        adUnitId = String(serverParameters[1])
        
        // Banner parameters
        if serverParameters.count == 3 {
            // Invalid parameter string, the loading will be cancelled with an error
            guard let bannerSize = Int(serverParameters[2]) else {
                completion(NSError(
                    domain: ErrorConstants.errorDomain,
                    code: ErrorConstants.errorCodeInvalidParameterString,
                    userInfo: [NSLocalizedDescriptionKey: "Invalid server parameter string: \(serverParameters)"]
                ))
                return
            }
            
            // Setting banner size
            if bannerSize == 0 {
                self.bannerSize = OguryAdsBannerSize.small_banner_320x50()
            } else if bannerSize == 1 {
                self.bannerSize = OguryAdsBannerSize.mpu_300x250()
            }
        }
        
        // Thumbnail parameters
        if serverParameters.count == 6 {
            // Invalid parameter string, the loading will be cancelled with an error
            guard let maxWidth = Int(serverParameters[2]),
                  let maxHeight = Int(serverParameters[3]),
                  let leftMargin = Int(serverParameters[4]),
                  let topMargin = Int(serverParameters[5]) else {
                      
                completion(NSError(
                    domain: ErrorConstants.errorDomain,
                    code: ErrorConstants.errorCodeInvalidParameterString,
                    userInfo: [NSLocalizedDescriptionKey: "Invalid server parameter string: \(serverParameters)"]
                ))
                return
            }
            
            // Setting Thumbnail size
            thumbnailSize = ThumbnailSize(maxWidth: maxWidth, maxHeight: maxHeight, leftMargin: leftMargin, topMargin: topMargin)
        }
        
        // Initializing Ogury SDK
        let oguryConfiguration = OguryConfigurationBuilder(assetKey: assetKey).build()
        Ogury.start(with: oguryConfiguration)
        
        // Ogury has been successfully initialized
        completion(nil)
    }
    
}
